### PR TITLE
Fix a regression in SSL_export_keying_material

### LINKED ISSUE #7712: Regression: Tor's use of SSL_export_keying_material() broken with TLS 1.3 in Openssl 1.1.1a
Hello!  Commit ec0c5f5693e39c5a013f81e6dd9dfd09ec65162d, "Reduce stack usage in tls13_hkdf_expand", breaks TLS1.3 when used with Tor.

The idea behind this commit, if I understand correctly, was to save stack space in `tls13_hkdf_expand` by lowering TLS13_MAX_LABEL_LEN, on the theory that TLS itself never uses any labels longer than 12 characters.

But the problem here is that the `tls13_hkdf_expand` function is also reachable from SSL_export_keying_material(), which is a public API.  As part of its handshake, Tor invokes this function with a 44-byte label.  This is no longer supported by OpenSSL 1.1.1a.

If we were to change Tor to use a shorter label, we'd break backward compatibility.

The simplest solution here would be to revert ec0c5f5693e39c5a013f81e6dd9dfd09ec65162d.

References:

#7297 is the original PR that introduced this problem for us.

https://trac.torproject.org/projects/tor/ticket/28616 is the Tor bug.

### PR DESCRIPTION
**Pull request by mattcaswell**
This reverts cbad0a61 which introduced a regression impacting Tor. This also fixes some issues in tls13_hkdf_expand() which impact the SSL_export_keying_material function for TLSv1.3. We add a test to check that we can use the maximum label length in TLSv1.3.

Fixes #7712

### CODE REVIEW

File: ssl/tls13_enc.c [hunk 1]
```diff
@@ -46,14 +40,28 @@ int tls13_hkdf_expand(SSL *s, const EVP_MD *md, const unsigned char *secret,
      * prefix and label + bytes for the label itself + 1 byte length of hash
      * + bytes for the hash itself
      */
-    unsigned char hkdflabel[sizeof(uint16_t) + sizeof(uint8_t) +
-                            + sizeof(label_prefix) + TLS13_MAX_LABEL_LEN
-                            + EVP_MAX_MD_SIZE];
+    static unsigned char hkdflabel[sizeof(uint16_t) + sizeof(uint8_t) +
```
Comment (t8m) [inline ssl/tls13_enc.c]: I suppose static here is a no-go as it would make the function non-reentrant. So it either needs to be kept on the stack or allocated.

File: ssl/tls13_enc.c [hunk 2]
```diff
@@ -29,14 +22,15 @@ static const unsigned char default_zeros[EVP_MAX_MD_SIZE];
...
+ * |fatal| is set. Returns 1 on success  0 on failure.
  */
 int tls13_hkdf_expand(SSL *s, const EVP_MD *md, const unsigned char *secret,
                              const unsigned char *label, size_t labellen,
                              const unsigned char *data, size_t datalen,
-                             unsigned char *out, size_t outlen)
+                             unsigned char *out, size_t outlen, int fatal)
 {
-    static const unsigned char label_prefix[] = "tls13 ";
+    const unsigned char label_prefix[] = "tls13 ";
```
Comment (t8m) [inline ssl/tls13_enc.c]: Why not keep the static here? Not that a reasonable compiler shouldn't be able to optimize it out anyway but...

Comment (mattcaswell) [reply ssl/tls13_enc.c]: Errr...I reverted the original commit and reimplemented some bits of it. I meant to retain the static - but it looks like I "missed" and added it to the wrong variable :-)

**Comment by mattcaswell** [comment]
Fixup commit pushed addressing feedback.

**Comment by mattcaswell** [comment]
Fixup commit pushed addressing travis failure.

**Comment by mattcaswell** [comment]
CIs are green. Pushed to master and 1.1.1.

**Comment by nmathewson** [comment]
Is it known when this fix will go into an openssl release? It is causing Tor to break on several downstream operating systems.

**Comment by mattcaswell** [comment]
We normally do releases on an "as needed" basis. This means either if a sufficiently serious security defect triggers it (moderate severity or above). If no security defect triggers a release then we typically do one anyway after 3-4 months from the previous release.

**Comment by jacmet** [comment]
@mattcaswell and for regressions like this?

**Comment by mattcaswell** [comment]
Well its not a security defect so it doesn't trigger a release itself. It will get rolled up in the next release whenever that ends up being. The last one was a little under 3 months ago, and we typically do one every 3-4 months so I would expect one "soon".
